### ISSUE TITLE
App ignores proxy settings

### ISSUE THREAD

**Issue by christian-muertz** [flutter/flutter#26359]
We need a way to access the system configured proxy.
In networks where a proxy is required to enter the internet, this is critical. 

Currently, I'm not able to use any flutter app in such networks where a proxy is configured in the system settings.

**Comment by zoechi** [comment]
Any reference to such functionality for native iOS/Android?

**Comment by christian-muertz** [comment]
Not familiar with native iOS/Android Development.
But I think `http.Client` should detect the global system proxy and use it so the developer does not need to think about it.

**Comment by zoechi** [comment]
If the proxy is configured in DHCP this might work but I think the OS would handle this then, not the app.

**Comment by christian-muertz** [comment]
I'm not that familiar with the topic but try using a flutter app that accesses the internet inside a network that requires the traffic to go to a proxy to enter the internet. For example in my school all the network traffic requires going through the schools proxy for logging etc. All apps manage to access the internet except flutter apps.

**Comment by christian-muertz** [comment]
Currently waiting for this issue to be resolved before releasing the app in the stores.

**Comment by zoechi** [comment]
Don't expect this to happen near-term. You'll need to find a workaround.

**Comment by zoechi** [comment]
In a quick search I found https://issuetracker.google.com/issues/36963864, https://en.wikipedia.org/wiki/Web_Proxy_Auto-Discovery_Protocol,
https://www.clearos.com/clearfoundation/social/community/setting-up-the-auto-proxy-redirection
but I don't have any special knowledge here

**Comment by christian-muertz** [comment]
It does not matter whether the proxy is detected automatically or is set manually in the system settings. The http client just ignores the system wide set proxy.

**Comment by zoechi** [comment]
Sorry, I misinterpreted the request.

**Comment by christian-muertz** [comment]
Sorry, title was misleading

**Comment by christian-muertz** [comment]
Might be related to #20376.

**Comment by zoechi** [comment]
https://docs.flutter.io/flutter/dart-io/HttpClient/findProxy.html might help (not tried)

**Comment by christian-muertz** [comment]
Issue is not to get the http client use the specified proxy. The issue is finding out the global proxy settings.

**Comment by zoechi** [comment]
Might be a duplicate of #20376 

See also about how to set a custom HttpClient that finds and applies proxy settings for every request:
https://github.com/flutter/flutter/issues/20376#issuecomment-439223882
https://github.com/flutter/flutter/issues/19588#issuecomment-406779390
https://stackoverflow.com/questions/54321077/best-way-to-set-default-header-for-all-request-in-flutter-http-request

**Comment by maks** [comment]
@zoechi Basically what this issue is asking for is not having to use a plugin as described in:  https://stackoverflow.com/a/50092642/85472

I guess the question is is it up to Http client to automatically use the host-system (android, ios, etc) configured proxy settings or is it up to the app developer to get that setting somehow and then use:
```
final proxyUrl = getHttpProxySomehowFromEmbeddingSystem();
class DebugHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(SecurityContext context) {
    return super.createHttpClient(context)
      ..findProxy = (uri) {
        return "PROXY $proxyUrl;";
      }
  }
}
```

**Comment by tghosth** [comment]
Hi @christian-muertz, did you get anywhere with this?

**Comment by duongtruong12** [comment]
Hi @christian-muertz , did you fix the problem, can you share experience please?

**Comment by SVyatoslavG** [comment]
I am having the same issue. Any news on this topic?

**Comment by christian-muertz** [comment]
> Hi @christian-muertz , did you fix the problem, can you share experience please?
> Hi @christian-muertz, did you get anywhere with this?

Nope, didn't keep working on it.

**Comment by tghosth** [comment]
This is sort of related: https://blog.nviso.be/2019/08/13/intercepting-traffic-from-android-flutter-applications/

**Comment by kasvith** [comment]
Having the same issue

**Comment by duongtruong12** [comment]
i had using dio plugin: [https://pub.dev/packages/dio](url)
And get proxySettings by native code and send back to flutter app to set on dio
`private Map getProxySettings() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            ProxyInfo defaultProxy = manager.getDefaultProxy();
            if (defaultProxy != null) {
                Map<String, String> map = new HashMap<>();
                map.put("port", Integer.toString(defaultProxy.getPort()));
                map.put("host", defaultProxy.getHost());
                return map;
            } else {
                return new HashMap();
            }
        }
        return new HashMap();
    }`

**Comment by sroddy** [comment]
We have been successful in passing the configurations by taking them from the platform code.
However I think that the native embedders should correctly set them inside the Dart environment so that the default `findProxy` implementation would correctly use the "expected" configuration/behaviour (I think that a system-wide set proxy configuration is expected from the dev to be the default when no custom options are set inside the http client)

cc @dnfield @chinmaygarde

**Comment by devmgs** [comment]
I created a class like
```
import 'dart:io';

class MyHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(SecurityContext context) {
    super.createHttpClient(context)
      ..findProxy = HttpClient.findProxyFromEnvironment;
  }
}
```

Now on my Apprun I call this before returning app

` HttpOverrides.global = new MyHttpOverrides();`

**Comment by RiteshBM** [comment]
> super.createHttpClient(context)

should be 

`return super.createHttpClient(context)`

?

**Comment by fralways** [comment]
App developers will usually miss this (unless encountered this issue before), thus newly released apps will not work when user has system configured proxy settings. Only after app starts getting bug reports, it can be pinpointed that system proxy is not set.

I would expect that the default behaviour is to use system settings, but if developer knows what he is doing, he could override this if needed, and not other way round.

**Comment by rhalff** [comment]
Here is a plugin which is able to get the system's proxy settings on android and ios:

https://github.com/kaivean/system_proxy

**Comment by REKURDennis** [comment]
Hi,

is there any plan to fix this issue in the future? Sadly for business apps this is a real problem, since many companies use some proxy configuration. And I fully agree with the previous comments, that using the system proxy should be the default behaviour.

**Comment by rajasekar-mu** [comment]
We also facing similar issue at corporate network on using android mobile app. tried with **findProxy** method, but did not get actual proxy information.
https://api.flutter.dev/flutter/dart-io/HttpClient/findProxy.html

@engine-flutter-autoroll, @jonahwilliams - many users facing CORP network issue, please fix as soon as possible.

**Comment by maks** [comment]
@rajasekar-mu you need to read the actually documentation you linked to, `findProxy`:

>Sets the function used to resolve the proxy server to be used 

*you* need to then provide a function that provides the proxy server to be used.

**Comment by rajasekar-mu** [comment]
yes @maks, i provided the function on initial loading. But its return localhost for proxy value, while debugging on corporate network. At the same time perfectly working internet users to return **DIRECT**. are you know any other way to solve this problem on fetching proxy information form corp network?

**Comment by jerryzhoujw** [comment]
> Here is a plugin which is able to get the system's proxy settings on android and ios:
> 
> https://github.com/kaivean/system_proxy

This one save my life.

**Comment by sebastianlyserena** [comment]
Has anyone solved the proxy for android 7 and up?

**Comment by ShokhrukhDeveloper** [comment]
HI

**Comment by devmgs** [comment]
I have successfully used the following method which works on android and ios.

https://pub.dev/packages/http_proxy

code is as follows

```
import 'package:http_proxy/http_proxy.dart';



void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  if (GetPlatform.isAndroid || GetPlatform.isIOS) {
    HttpProxy proxy = await HttpProxy.createHttpProxy();
    HttpOverrides.global = proxy;
  }
.........
```

**Comment by komaxx** [comment]
@devmgs 
I think you might want to re-consider your solution!

`http_proxy` contains the following lines:

```dart
var client = super.createHttpClient(context);
    client.badCertificateCallback =
            (X509Certificate cert, String host, int port) {
      return true;
    };
    return client;
```

This effectively turns off all certificate checking. With it, your app is wide open to a wide range of attacks :(

Granted, I don't have a better solution right now but I think we should not be happy with this workaround..

**Comment by NTA-trongpq** [comment]
> > Here is a plugin which is able to get the system's proxy settings on android and ios:
> > https://github.com/kaivean/system_proxy
> 
> This one save my life.

Does this working when use proxy with authentication (username & password)

**Comment by hariprasad47** [comment]
> I have successfully used the following method which works on android and ios.
> 
> https://pub.dev/packages/http_proxy
> 
> code is as follows
> 
> ```
> import 'package:http_proxy/http_proxy.dart';
> 
> 
> 
> void main() async {
>   WidgetsFlutterBinding.ensureInitialized();
> 
>   if (GetPlatform.isAndroid || GetPlatform.isIOS) {
>     HttpProxy proxy = await HttpProxy.createHttpProxy();
>     HttpOverrides.global = proxy;
>   }
> .........
> ```
This works on Android but I am still unable to make it work on iOS device does anyone have a solution to make flutter proxy aware for iOS devices.

**Comment by sherry0429** [comment]
I create a pr https://github.com/flutter/flutter/pull/122170, 
it modify fluter_windows.cpp to make windows application request through system proxy.
it now closed, becasue want to keep these app templates as minimal as possible.

but if anyone need way to send request by proxy you cloud checkout this pr

**Comment by Rajala1404** [comment]
> > I have successfully used the following method which works on android and ios.
> > https://pub.dev/packages/http_proxy
> > code is as follows
> > ```
> > import 'package:http_proxy/http_proxy.dart';
> > 
> > 
> > 
> > void main() async {
> >   WidgetsFlutterBinding.ensureInitialized();
> > 
> >   if (GetPlatform.isAndroid || GetPlatform.isIOS) {
> >     HttpProxy proxy = await HttpProxy.createHttpProxy();
> >     HttpOverrides.global = proxy;
> >   }
> > .........
> > ```
> 
> This works on Android but I am still unable to make it work on iOS device does anyone have a solution to make flutter proxy aware for iOS devices.

I actually made an Fork oh http_proxy that supports iOS and PAC
https://github.com/lanis-mobile/http_proxy

**Comment by UnluckyY1** [comment]
Wow, it’s been 5 years and this issue has never been addressed.

If you’re facing the same issue , I found a workaround that works for Android, iOS, and Windows using the system_proxy plugin. This solution allows your app to dynamically retrieve the system’s proxy settings and apply them globally. It works with both Dio and http, so you can use it seamlessly for network requests.


1.	Add the Dependency to pubspec.yaml:
```yaml
dependencies:
  system_proxy:
    git:
      url: https://github.com/UnluckyY1/system_proxy_windows.git
      ref: 10be4f7
```

Step 2: Create custom_proxy.dart for handling proxy settings

```dart
import 'dart:io';

/// Allows you to set and enable a proxy for your app
class CustomProxy {
  /// A string representing an IP address for the proxy server
  final String? ipAddress;

  /// The port number for the proxy server
  /// Can be null if port is default.
  final int? port;

  /// Set this to true
  /// - Warning: Setting this to true in production apps can be dangerous. Use with care!
  bool allowBadCertificates;

  /// Initializer
  CustomProxy({required this.ipAddress, this.port, this.allowBadCertificates = false});

  /// Initializer from string
  /// Note: Uses static method, rather than named init to allow final properties.
  static CustomProxy? fromString({required String proxy}) {
    // Check if valid
    if (proxy == '') {
      assert(false, 'Proxy string passed to CustomProxy.fromString() is invalid.');
      return null;
    }

    // Build and return
    final proxyParts = proxy.split(':');
    final ipAddress = proxyParts[0];
    final port = proxyParts.isNotEmpty ? int.tryParse(proxyParts[1]) : null;
    return CustomProxy(
      ipAddress: ipAddress,
      port: port,
    );
  }

  /// Enable the proxy
  void enable() {
    HttpOverrides.global = CustomProxyHttpOverride.withProxy(toString());
  }

  /// Disable the proxy
  void disable() {
    HttpOverrides.global = null;
  }

  @override
  String toString() {
    String proxy = ipAddress ?? '';
    if (port != null) {
      proxy += ':$port';
    }
    return proxy;
  }
}

/// This class overrides the global proxy settings.
class CustomProxyHttpOverride extends HttpOverrides {
  /// The entire proxy server
  /// Format: "localhost:8888"
  final String proxyString;

  /// Set this to true
  /// - Warning: Setting this to true in production apps can be dangerous. Use with care!
  final bool allowBadCertificates;

  /// Initializer
  CustomProxyHttpOverride.withProxy(
    this.proxyString, {
    this.allowBadCertificates = false,
  });

  /// Override HTTP client creation
  @override
  HttpClient createHttpClient(SecurityContext? context) {
    return super.createHttpClient(context)
      ..findProxy = (uri) {
        assert(
          proxyString.isNotEmpty,
          'You must set a valid proxy if you enable it!',
        );
        return 'PROXY $proxyString;';
      }
      ..badCertificateCallback = allowBadCertificates ? (X509Certificate cert, String host, int port) => true : null;
  }
}
```

Step 3: Create a method to get the system proxy settings



```dart
/// Retrieves the system proxy settings
Future<CustomProxy?> getSystemProxy() async {
  // Retrieve proxy settings from the system
  final proxy = await SystemProxy.getProxySettings();
  
  if (proxy != null && proxy['host'] != null && proxy['port'] != null) {
    // Return CustomProxy with system-provided settings
    return CustomProxy(
      ipAddress: proxy['host']!,
      port: proxy['port'],
    );
  }
  // Return null if no proxy is configured
  return null;
}
```

Step 4: Using the proxy in your app


```dart
void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  final proxy = await getSystemProxy();
  if (proxy != null) {
    proxy.enable();  // Apply proxy settings globally
    print('Proxy enabled: ${proxy.toString()}');
  } else {
    print('No system proxy found.');
  }

  runApp(MyApp());
}


```
**Edit** : As of now, for Android, iOS, and macOS you can use nativeAdapter (such as native_dio_adapter with Dio) or rely on http, which uses the native implementation for each platform. However, WinHTTP has not yet been added, so you will still need this for Flutter apps running on Windows.

**Comment by UnluckyY1** [comment]
@dnfield   @chinmaygarde @kf6gpe 

I was very surprised when one of our clients reported that our app was ignoring system proxy settings while running on Windows. After investigating, I realized this is a long-standing issue with Flutter/Dart, where system-wide proxy configurations are not respected by default.

This behavior is unexpected and problematic, especially for enterprise environments where system proxy settings are commonly used for security, monitoring, or restricted network configurations.

Expected Behavior

When no custom proxy settings are specified in the HTTP client, the app should default to using the system-wide proxy configuration. This is standard behavior for many platforms and frameworks and is what developers and clients naturally expect.

Suggested Solution

The root of the issue seems to lie in how native embedders handle proxy settings. I believe the native embedders should correctly pass the system proxy configurations to the Dart environment, allowing the default findProxy implementation to use the expected system settings. This would make apps behave consistently with other software running on the same system.

It has been 5 years, and this basic functionality is still not implemented. This is a fundamental requirement that should be prioritized.

**Comment by lotfi-a** [comment]
@UnluckyY1 7 years now ...

**Comment by devmgs** [comment]
Just a work around to detect proxy use plugin in pubspec
http_proxy_override: ^1.0.1


in main.dart
========


```
void main() async {

WidgetsFlutterBinding.ensureInitialized();
 
if (Platform.isAndroid \|\| Platform.isIOS) {
HttpProxyOverride httpProxyOverride =  await HttpProxyOverride.createHttpProxy();
HttpOverrides.global = httpProxyOverride;
}


/// You APP code follows

}
```

**Comment by devmgs** [comment]
Till it is not implemented use https://pub.dev/packages/detect_proxy_setting

**Comment by devmgs** [comment]
> User can change proxy settings while app is running. Setting up proxy settings at start, and not listening to changes is also not the best.

Agree.

Actually its issue is within dart on which flutter is based. Till it is resolved we can try to achieve this using connectivity_plus.

Something like Connectivity Plus detects changes in network and connectivity if it identifies discinnection the proxy can be set by detect_proxy_setting to verify if it now connects we can define a cooldown period / provision in UI to update proxy from environment. I know this is cumbersome but till then we need to do apply this workaround so our applications don't face issues.

**Comment by vdemcak** [comment]
Any official response to this? It's kind of insane, that an issue can be open for almost 7 years with this level of attention and still not be addressed.

**Comment by UnluckyY1** [comment]
I had a discussion with Slava Egorov (Vyacheslav Egorov) on X about this around 6 months ago.
From that conversation and my understanding , it was clear that the Dart team doesn’t plan to support this on Windows. The reason is that this isn’t considered a Dart language issue 

The recommended approach for Windows would be to implement a WinHTTP-based plugin, similar to:

- cupertino_http for iOS/macOS
- cronet_http for Android

A new win_http plugin could wrap the native WinHTTP API (via Dart FFI), ensuring consistent behavior without relying on the weak dartio client.
